result/result.py

```diff
@@ -1,211 +1,245 @@
+        """
+        return op(self._value)
+
+    def map_err(self, op: Callable[[E], F]) -> 'Result[T, F]':
+        """
+        The contained result is `Ok`, so return `Ok` with the original value
+        """
+        return cast(Result[T, F], self)
+
+
+class Err(Generic[E]):
+    """
+    A value that signifies failure and which stores arbitrary data for the error.
+    """
+
+    def __init__(self, value: E) -> None:
+        self._value = value
+
+    def __repr__(self) -> str:
+        return "Err({})".format(repr(self._value))
+
+    def __eq__(self, other: Any) -> bool:
+        return isinstance(other, Err) and self.value == other.value
+
+    def __ne__(self, other: Any) -> bool:
+        return not (self == other)
+
+    def __hash__(self) -> int:
+        return hash((False, self._value))
 
     def is_ok(self) -> bool:
-        return self._is_ok
+        return False
 
     def is_err(self) -> bool:
-        return not self._is_ok
+        return True
 
-    def ok(self) -> Optional[T]:
+    def ok(self) -> None:
         """
-        Return the value if it is an `Ok` type. Return `None` if it is an
-        `Err`.
+        Return `None`.
         """
-        return cast(T, self._value) if self.is_ok() else None
+        return None
 
-    def err(self) -> Optional[E]:
+    def err(self) -> E:
         """
-        Return the error if this is an `Err` type. Return `None` otherwise.
+        Return the error.
         """
-        return cast(E, self._value) if self.is_err() else None
+        return self._value
 
     @property
-    def value(self) -> Union[T, E]:
+    def value(self) -> E:
         """
-        Return the inner value. This might be either the ok or the error type.
+        Return the inner value.
         """
         return self._value
 
-    def expect(self, message: str) -> T:
+    def expect(self, message: str) -> NoReturn:
         """
-        Return the value if it is an `Ok` type. Raises an `UnwrapError` if it
-        is an `Err`.
+        Raises an `UnwrapError`.
         """
-        if self._is_ok:
-            return cast(T, self._value)
-        else:
-            raise UnwrapError(message)
+        raise UnwrapError(message)
 
-    def expect_err(self, message: str) -> E:
+    def expect_err(self, _message: str) -> E:
         """
-        Return the value if it is an `Err` type. Raises an `UnwrapError` if it
-        is `Ok`.
+        Return the inner value
         """
-        if self._is_ok:
-            raise UnwrapError(message)
-        return cast(E, self._value)
+        return self._value
 
-    def unwrap(self) -> T:
+    def unwrap(self) -> NoReturn:
         """
-        Return the value if it is an `Ok` type. Raises an `UnwrapError` if it
-        is an `Err`.
+        Raises an `UnwrapError`.
         """
-        return self.expect("Called `Result.unwrap()` on an `Err` value")
+        raise UnwrapError("Called `Result.unwrap()` on an `Err` value")
```

Comment:
Should we keep `Result.` in the error messages even if it's not a real type anymore?

(It might or might not be a good idea to keep it. I'm not sure.)

Reply:
I would keep it, while `Result` isn't a real object you can still reason about the union between `Ok` and `Err`, and that union has the method `unwrap`.